## Adedayo Adegoke Portfolio

**Data Cleaning Tasks:**

Some of the variables in the ports.csv datasets are not presented in a format suitable for
the calculation of statistics in their original format.

**1. (a) The dataset ports.csv contains missing values.**
Create and test a Python script that will identify, and either remove rows that
contain missing data from your DataFrame, or better, fill any missing values
with suitable values.

  **(b) The dataset ports.csv contains duplicate rows.**
Create and test a Python script that will identify and remove any duplicate
rows from your DataFrame.

**2. (a) The rain column is in text format due to the mm unit in some of the rows.**
Create and test a Python script that will convert the rain column into
numerical format, e.g. "3.1mm” should become 3.1

 **(b) The labour column has a mixture of formats.**
 In 2021 the values used are ‘0’ and ‘1’, whilst in 2022 the values used are ‘striking’ and ‘working’. However
for the purposes of analysis ‘0’ is equivalent to ‘striking’ and ‘1’ to ‘working’

### Create and test a Python script that will convert the labour column into a consistent format containing only the values working or striking.

This initial section focuses on preparing the 'ports.csv' dataset for analysis. It addresses the need to reformat certain variables that are not initially suitable for statistical calculations.

In the context of dataset analysis, Renear, A.H., Sacchi, S. and Wickett, K.M. (2010) in their work "Definitions of dataset in the scientific and technical literature" (Proceedings of the American Society for Information Science and Technology, 47(1), pp.1–4) underscore the significance of clearly defining variables for accurate interpretation and analysis. The dataset for this project is describe below:

Time (Target Variable): The primary variable of interest, representing the duration we aim to predict.

TEU (Total Equivalent Units): This metric quantifies a ship's cargo capacity, vital for understanding loading efficiency.

Loadratio: A ratio depicting the actual versus potential cargo load of a ship, indicating operational efficiency.

Gear: This likely refers to the equipment utilized in cargo handling, a factor influencing operational time.

Wind: Wind speed is included as it can significantly affect loading and unloading operations.

Rain: Precipitation is a critical factor, potentially impacting loading/unloading efficiency.

Weather: This categorical variable describes overall weather conditions, providing context for operational delays or efficiencies.

onSchedule: Indicates whether a ship adheres to its schedule, a factor that could lead to prioritized handling.

Labour: Reflects workforce status, directly correlating with operational time and efficiency.

Origin: The ship's origin could influence handling time due to varying cargo types and loading procedures.

Understanding these variables is crucial for comprehensively analysing the dataset and deriving meaningful conclusions, aligning with Renear et al.’s emphasis on the importance of precise variable definition in data analysis.

## Task 1

To identify and replace any cells containing missing data in the DataFrame, I will follow these steps:

Load the data from the CSV file into a pandas DataFrame.
Examine the data to identify which columns are numerical and which are categorical.
For numerical columns: Calculate the mean of the column. Replace missing values with the mean of that column.
For categorical columns: Determine the mode (most common value) of the column. Replace missing values with the mode of that column.

In [6]:
# Loading necessary library
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np 
import statistics as stat


# Load the data
ports = pd.read_csv("ports.csv")

# Display the first few rows of the DataFrame to understand its structure
print(ports.shape)
ports.head()

(1006, 13)


time   TEU  loadratio  gear  wind    rain     weather onSchedule  \
0  263.2  8645       85.4   0.0   3.9     8.4  disruptive        Yes   
1  177.9  2106       80.8   1.0   7.5   7.7mm  disruptive        Yes   
2  269.3  7872       90.4   3.0   6.9  19.3mm  disruptive        Yes   
3  122.2  1938       72.2   1.0   8.9     3.8      normal        Yes   
4  165.0  1669       80.0   4.0   2.0     0.1      normal         No   

     labour            origin delay        date       port  
0   working  Intercontinental    No  2022-11-01  Abermouth  
1         1  Intercontinental   Yes  2021-09-09  Abermouth  
2         1  Intercontinental    No  2021-08-05  Abermouth  
3   working                UK    No  2022-11-03  Abermouth  
4  striking                UK   Yes  2022-02-01  Abermouth

In [2]:
#Comprehensive Statistical Overview of Ports Dataset
ports.describe()

time           TEU    loadratio        gear        wind
count  1006.000000   1006.000000  1006.000000  994.000000  985.000000
mean    242.009742   6175.164016    80.049503    1.816901    5.702234
std      85.278670   3615.173895     9.739284    1.474413    4.157579
min      49.400000    205.000000    42.500000    0.000000    0.100000
25%     177.900000   2930.250000    73.900000    1.000000    2.600000
50%     231.400000   6274.000000    80.100000    2.000000    4.800000
75%     295.475000   9420.750000    86.800000    3.000000    8.100000
max     582.000000  11999.000000   100.000000    6.000000   25.500000

Dataset consist the 1006 rows with the 13 columns consisting of both numberical and categorical variables. The 'rain' column, which is numerical but erroneously read as an object due to the presence of strings like "7.7mm".

The labour column has inconsistent mixture of data types. This is why the summary statistics could not be described for them using the describe() keyword.

Next, I will identify missing data in the DataFrame and replace with their mean value.

In [4]:
# Check for missing values in each column
missing_values = ports.isnull().sum()
print("Missing values in each column:")
print(missing_values)

# Identify columns with missing values
columns_with_missing_values = missing_values[missing_values > 0].index
print("Columns with missing values:")
print(columns_with_missing_values)

# Print the shape of the dataset
print("Shape of the dataset:")
print(ports.shape)

Missing values in each column:
time           0
TEU            0
loadratio      0
gear          12
wind          21
rain           0
weather        0
onSchedule     0
labour         0
origin         0
delay          0
date           0
port           0
dtype: int64
Columns with missing values:
Index(['gear', 'wind'], dtype='object')
Shape of the dataset:
(1006, 13)


Initial state of the dataset for comparison: The ports dataset, encompassing 1006 entries and 13 attributes, reveals missing values exclusively in the 'gear' and 'wind' columns, 12 and 21 instances, respectively. This indicates a need for careful handling or imputation in these areas, ensuring robust analysis while maintaining the dataset's comprehensive scope.

Renaming the DataFrame to ports_replacemissing is a strategic decision in data analysis, particularly when dealing with multiple DataFrames or stages in data processing. It is beneficial in the follwing ways:

Clarity and Context
Avoiding Conflicts
Facilitating Modular Code
Preparing for Data Cleaning or Transformation Tasks

In [8]:
# Loading necessary library
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np 
import statistics as stat

# Load the data
ports_replacemissing = pd.read_csv("ports.csv")

# Display the first few rows of the DataFrame to understand its structure
print(ports_replacemissing.shape)
ports_replacemissing.head()

(1006, 13)


time   TEU  loadratio  gear  wind    rain     weather onSchedule  \
0  263.2  8645       85.4   0.0   3.9     8.4  disruptive        Yes   
1  177.9  2106       80.8   1.0   7.5   7.7mm  disruptive        Yes   
2  269.3  7872       90.4   3.0   6.9  19.3mm  disruptive        Yes   
3  122.2  1938       72.2   1.0   8.9     3.8      normal        Yes   
4  165.0  1669       80.0   4.0   2.0     0.1      normal         No   

     labour            origin delay        date       port  
0   working  Intercontinental    No  2022-11-01  Abermouth  
1         1  Intercontinental   Yes  2021-09-09  Abermouth  
2         1  Intercontinental    No  2021-08-05  Abermouth  
3   working                UK    No  2022-11-03  Abermouth  
4  striking                UK   Yes  2022-02-01  Abermouth

Deliberately renaming the DataFrame to ports_replacemissing is a good practice in data analysis, as it provides clarity, prevents potential conflicts, and sets the stage for specific data processing tasks.

## Data cleansing and wrangling

Various methods have been used historically for missing values imputation. Dadi, K et al. (2021) used mean for imputation as it maintains the column's overall distribution and it is widely used. Therefore missing values for Gear and Wind shall be replaced their mean.

In [9]:
# replace missing values with mean for each numeric column
for col in ports_replacemissing.select_dtypes(include=['float64', 'int64']):
    ports_replacemissing[col].fillna(ports_replacemissing[col].mean(), inplace=True)

# Print the dataset after replacing missing values
print("\nDataset after replacing missing values with their mean:")
ports_replacemissing.head()


Dataset after replacing missing values with their mean:


time   TEU  loadratio  gear  wind    rain     weather onSchedule  \
0  263.2  8645       85.4   0.0   3.9     8.4  disruptive        Yes   
1  177.9  2106       80.8   1.0   7.5   7.7mm  disruptive        Yes   
2  269.3  7872       90.4   3.0   6.9  19.3mm  disruptive        Yes   
3  122.2  1938       72.2   1.0   8.9     3.8      normal        Yes   
4  165.0  1669       80.0   4.0   2.0     0.1      normal         No   

     labour            origin delay        date       port  
0   working  Intercontinental    No  2022-11-01  Abermouth  
1         1  Intercontinental   Yes  2021-09-09  Abermouth  
2         1  Intercontinental    No  2021-08-05  Abermouth  
3   working                UK    No  2022-11-03  Abermouth  
4  striking                UK   Yes  2022-02-01  Abermouth

In [10]:
# Check for missing values in each column
missing_values = ports_replacemissing.isnull().sum()
print("Missing values in each column:")
print(missing_values)

# Remove rows with missing data
ports_replacemissing = ports_replacemissing.dropna()

# Print the shape of the DataFrame after removing rows with missing data
print("Shape of the dataset after removing rows with missing data:")
print(ports_replacemissing.shape)

Missing values in each column:
time          0
TEU           0
loadratio     0
gear          0
wind          0
rain          0
weather       0
onSchedule    0
labour        0
origin        0
delay         0
date          0
port          0
dtype: int64
Shape of the dataset after removing rows with missing data:
(1006, 13)


## QUESTION 1(b)

To identify and remove any duplicate rows from the DataFrame, I will:

Use the duplicated() method to identify any duplicate rows, marking them as True if they are duplicates.
Use the drop_duplicates() method to remove the duplicate rows.
I will then display the number of duplicate rows identified and removed.

In [12]:
#Identifying duplicate rows
duplicate_rows = ports_replacemissing.duplicated()
num_duplicates = duplicate_rows.sum()

# Displaying the number of duplicate rows
print(num_duplicates)

6


The dataset contains 6 duplicate rows. I will now remove these duplicates and confirm their removal.

Damodaram (2022) and Steorts (2023) emphasise the importance of removing duplicates in data management and analysis, highlighting its role in maintaining data integrity and quality.

The below code identifies and displays the number of duplicate rows in the 'ports' DataFrame.

In [13]:
# Removing duplicate rows
ports_replacemissing = ports_replacemissing.drop_duplicates()

# Verifying if duplicates are removed
remaining_duplicates = ports_replacemissing.duplicated().sum()

# Display the result
remaining_duplicates, ports_replacemissing.shape
print(remaining_duplicates, ports_replacemissing.shape)

0 (1000, 13)


I have successfully removed six duplicates rows from our dataset. As a result, the dataset now comprises 1000 rows and 13 columns.

## QUESTION 2 (a) Convert the rain column from text to numerical format.¶

I will strip the "mm" from the values and convert the column to a float type.

In [14]:
 # Strip "mm" from the 'rain' column and convert to float
# ports_replacemissing['rain'] = ports_replacemissing['rain'].str.replace('mm', '').astype(float)

# # Display the modified data
# ports_replacemissing.head()

# Check if 'rain' column exists in the DataFrame, then strip "mm" and convert to float
if 'rain' in ports_replacemissing.columns:
    ports_replacemissing['rain'] = ports_replacemissing['rain'].str.replace('mm', '').astype(float)
else:
    print("The 'rain' column does not exist in the DataFrame.")

# Display the modified data
ports_replacemissing.head()

time   TEU  loadratio  gear  wind  rain     weather onSchedule    labour  \
0  263.2  8645       85.4   0.0   3.9   8.4  disruptive        Yes   working   
1  177.9  2106       80.8   1.0   7.5   7.7  disruptive        Yes         1   
2  269.3  7872       90.4   3.0   6.9  19.3  disruptive        Yes         1   
3  122.2  1938       72.2   1.0   8.9   3.8      normal        Yes   working   
4  165.0  1669       80.0   4.0   2.0   0.1      normal         No  striking   

             origin delay        date       port  
0  Intercontinental    No  2022-11-01  Abermouth  
1  Intercontinental   Yes  2021-09-09  Abermouth  
2  Intercontinental    No  2021-08-05  Abermouth  
3                UK    No  2022-11-03  Abermouth  
4                UK   Yes  2022-02-01  Abermouth

The 'rain' column's "mm" values have been stripped, converting the data to a numerical type. It is crucial to format data appropriately for precise and efficient processing in data analysis. Take the 'rain' column as an example: changing values from a string format like "3.1mm" to a numerical one such as 3.1 is essential for quantitative analysis.

Using numerical data allows for various statistical calculations, such as computing averages, examining correlations, and performing regression analyses, which are complex with string data.

## QUESTION 2 (b) Standardise the labour column to have consistent formatting.

I intend to unify the 'labour' column by substituting '0' with 'striking' and '1' with 'working', ensuring consistency throughout the column.

In [15]:
# Replace '0' with 'striking' and '1' with 'working' in the 'labour' column
ports_replacemissing['labour'] = ports_replacemissing['labour'].replace({'0': 'striking', '1': 'working'})

# Display the modified data
ports_replacemissing.head()

time   TEU  loadratio  gear  wind  rain     weather onSchedule    labour  \
0  263.2  8645       85.4   0.0   3.9   8.4  disruptive        Yes   working   
1  177.9  2106       80.8   1.0   7.5   7.7  disruptive        Yes   working   
2  269.3  7872       90.4   3.0   6.9  19.3  disruptive        Yes   working   
3  122.2  1938       72.2   1.0   8.9   3.8      normal        Yes   working   
4  165.0  1669       80.0   4.0   2.0   0.1      normal         No  striking   

             origin delay        date       port  
0  Intercontinental    No  2022-11-01  Abermouth  
1  Intercontinental   Yes  2021-09-09  Abermouth  
2  Intercontinental    No  2021-08-05  Abermouth  
3                UK    No  2022-11-03  Abermouth  
4                UK   Yes  2022-02-01  Abermouth

I have made the 'labour' column consistent by replacing '0' with 'striking' and '1' with 'working'. This standardisation of categorical data, demonstrated by the transformation of '0' and '1' to 'striking' and 'working', enhances clarity and uniformity in the dataset. It not only aids in category-based analysis but also improves the readability of the data.

This process of data cleaning and standardisation is crucial, as emphasised by Steorts (2023) in "A Primer on the Data Cleaning Pipeline" and Zacks, Kenett, and Gedeck (2023) in "Modern Statistics: A Computer-Based Approach with Python Solutions." These authors underline the importance of thorough data preparation for achieving accurate and insightful analysis, highlighting the fundamental role of data preprocessing in statistics.